### ISSUE TITLE
remote_directory produces inconsistent dir permissions

### ISSUE THREAD

**Issue by andy-dufour** [chef/chef#3942]
Moved from:

https://tickets.opscode.com/browse/CHEF-5080

Dir to be synced:
```
./broken1/aaa
./broken1/dir/file
./broken2/dir/file
./broken2/xxx
```
with:
```
remote_directory "/tmp/remote_something" do
  source "something"
  files_owner "fileowner"
  files_group "fileowner"
  files_mode 0700
  owner "dirowner"
  group "dirowner"
  mode 0750
end
```
Produces:
```
/tmp/remote_something/
drwxr-x--- 6 dirowner root 4.0K Feb 25 21:35 .
drwxrwxrwt 4 root     root 4.0K Feb 25 21:35 ..
drwxr-xr-x 3 root     root 4.0K Feb 25 21:35 broken1
drwxr-x--- 3 dirowner root 4.0K Feb 25 21:35 broken2

/tmp/remote_something/broken1:
drwxr-xr-x 3 root      root 4.0K Feb 25 21:35 .
drwxr-x--- 6 dirowner  root 4.0K Feb 25 21:35 ..
-rwx------ 1 fileowner root    0 Feb 25 21:35 aaa
drwxr-x--- 2 dirowner  root 4.0K Feb 25 21:35 dir

/tmp/remote_something/broken2:
drwxr-x--- 3 dirowner  root 4.0K Feb 25 21:35 .
drwxr-x--- 6 dirowner  root 4.0K Feb 25 21:35 ..
drwxr-x--- 2 dirowner  root 4.0K Feb 25 21:35 dir
-rwx------ 1 fileowner root    0 Feb 25 21:35 xxx
```

Owner of `/tmp/remote_something/broken1` is wrong (not managed), while `/tmp/remote_something/broken2` gets expected owner (managed).

Instead of doing alphabetical ordering on the paths while creating files, which results in undefined result, it should traverse breadth-first.

**Comment by andy-dufour** [comment]
PR open for this: https://github.com/chef/chef/pull/3944

**Comment by jaymzh** [comment]
YAY!!!! :+1:

**Comment by jkeiser** [comment]
Fixed with #3944
